### ISSUE TITLE
bug: Youtube with spoof-wifi-connection patch crashes immediately

### ISSUE THREAD

**Issue by elim-garak-dev** [revanced/revanced-patches#2177]
### Type

Error at runtime

### Bug description

Patched youtube with spoof-wifi-connection patch is crashing on start. Without this patch youtube starts fine.
With recommended patches is that same. I have been registering this behavior for several weeks.

### Steps to reproduce

Patch youtube 18.16.37 with revanced-cli commands.

java -jar revanced-cli-2.21.2-dev.2-all.jar \
 -a com.google.android.youtube_18.16.37.apk \
 -m revanced-integrations-0.108.0-dev.17.apk \
 -o YouTube-ReVanced.apk \
 -b revanced-patches-2.174.0-dev.34.jar \
 -i custom-video-buffer \
 -i enable-android-debugging \
 -i export-all-activities \
 -i hide-player-overlay \
 -i predictive-back-gesture \
 -i remove-screenshot-restriction \
 -i remove-player-button-background \
 -e spoof-wifi-connection \
 --experimental

install it as non-root.

### Relevant log output

```shell
05-17 20:20:21.386  9738  9738 E AndroidRuntime: FATAL EXCEPTION: main
05-17 20:20:21.386  9738  9738 E AndroidRuntime: Process: app.revanced.android.youtube, PID: 9738
05-17 20:20:21.386  9738  9738 E AndroidRuntime: java.lang.VerifyError: Verifier rejected class jti: java.lang.Boolean jti.i(alit, long) failed to verify: java.lang.Boolean jti.i(alit, long): [0x93] register v6 has type Long (Low Half) but expected Precise Reference: java.lang.String (declaration of 'jti' appears in base.apk!classes3.dex)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at jst.s(PG:1)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at fkk.n(PG:88)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at fkk.a(PG:51)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at ataa.a(PG:1)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at jxq.o(PG:1)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at fkk.n(PG:89)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at fkk.a(PG:51)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at ataa.a(PG:1)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at fkk.j(PG:96)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at fkk.a(PG:55)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at ataa.a(PG:1)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at fkk.j(PG:97)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at fkk.a(PG:55)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at ataa.a(PG:1)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at aszz.a(PG:2)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at fkk.j(PG:72)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at fkk.a(PG:55)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at ataa.a(PG:1)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at aszz.a(PG:2)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at com.google.android.apps.youtube.app.watchwhile.WatchWhileActivity.o(PG:15)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at ija.a(PG:28)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at qq.onCreate(PG:3)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at bt.onCreate(PG:1)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at fvs.onCreate(PG:8)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at mjq.onCreate(PG:1)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at app.revanced.integrations.swipecontrols.SwipeControlsHostActivity.onCreate(SwipeControlsHostActivity.kt:72)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at com.google.android.apps.youtube.app.watchwhile.WatchWhileActivity.onCreate(PG:4)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at android.app.Activity.performCreate(Activity.java:8108)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at android.app.Activity.performCreate(Activity.java:8092)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1309)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3481)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3683)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2132)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:250)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7848)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
05-17 20:20:21.386  9738  9738 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:958)
```

### Additional context

Android 11, moto g9plus, vanced microg instaled, stock youtube is disabled.

### Acknowledgements

- [X] I have searched the existing issues and this is a new and no duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I filled out all of the requested information in this issue properly.

**Comment by elim-garak-dev** [comment]
I do this with -e spoof-wifi-connection.
I will wait.

**Comment by SodaWithoutSparkles** [comment]
It is excluded by default.

**Comment by 0xrxL** [comment]
```
[0x93] register v6 has type Long (Low Half) but expected Precise Reference: java.lang.String
```

Wrong register 😋

**Comment by Linus789** [comment]
Patching with `--exclusive -i spoof-wifi-connection -i client-spoof -i vanced-microg-support` works fine. Therefore, at least one other patch has to be involved. After some testing, the other involved patch seems to be `hide-info-cards`.

**Comment by oSumAtrIX** [comment]
That is odd, have you tried patching with all default patches but excluding `hide-info-cards` to confirm it is really that one?

**Comment by Linus789** [comment]
Yes, and I find it odd, too. I used
<details>
<pre>
  --exclusive \
  -i spoof-wifi-connection \
  -i vanced-microg-support \
  -i client-spoof \
  -i always-autorepeat \
  -i comments \
  -i copy-video-url \
  -i custom-branding \
  -i custom-video-buffer \
  -i custom-video-speed \
  -i disable-auto-captions \
  -i disable-fullscreen-panels \
  -i disable-player-popup-panels \
  -i disable-shorts-on-startup \
  -i disable-zoom-haptics \
  -i downloads \
  -i enable-android-debugging \
  -i enable-debugging \
  -i export-all-activities \
  -i general-ads \
  -i hdr-auto-brightness \
  -i hide-album-cards \
  -i hide-artist-card \
  -i hide-autoplay-button \
  -i hide-breaking-news-shelf \
  -i hide-captions-button \
  -i hide-cast-button \
  -i hide-crowdfunding-box \
  -i hide-email-address \
  -i hide-endscreen-cards \
  -i hide-floating-microphone-button \
  -i hide-get-premium \
  -i hide-player-buttons \
  -i hide-player-overlay \
  -i hide-seekbar \
  -i hide-timestamp \
  -i hide-video-action-buttons \
  -i hide-watch-in-vr \
  -i hide-watermark \
  -i minimized-playback \
  -i navigation-buttons \
  -i old-quality-layout \
  -i open-links-externally \
  -i predictive-back-gesture \
  -i premium-heading \
  -i remember-playback-speed \
  -i remember-video-quality \
  -i remove-player-button-background \
  -i remove-screenshot-restriction \
  -i return-youtube-dislike \
  -i seekbar-tapping \
  -i sponsorblock \
  -i spoof-app-version \
  -i swipe-controls \
  -i tablet-mini-player \
  -i theme \
  -i video-ads \
  -i wide-searchbar \
</pre>
</details>
and
<details>
<pre>
  --exclusive \
  -i spoof-wifi-connection \
  -i vanced-microg-support \
  -i hide-info-cards \
</pre>
</details>

and both crash.

**Comment by oSumAtrIX** [comment]
This indicates that the patch interferes with other patches. My best theory:

Patches that modify classes create a mutable copy of the class, but this class is not used to resolve fingerprints. Instead, the original classes are used. If a patch removes all method instructions, another fingerprint may still resolve to that method.

There exists `MethodFingerprint.mutableMethod`, which looks up if there exists a mutable copy already or creates one, and `MethodFingerprint.method` that returns the original unmodified method on which the fingerprint has been resolved. Now, since patches use the `mutableMethod` property, our method has no more instructions in our example because another patch has modified it.

TLDR: Shared mutability between patches.

This issue may be solved in ReVanced Patcher if the way fingerprints are resolved turns out incorrect. Alternatively, patches can manually resolve method fingerprints on (mutable) methods (because they implement `Method`).

Please let me know if something is unclear.

**Comment by Linus789** [comment]
Thanks, it seems like you nailed the issue! I didn’t even know that it uses the original method to resolve the fingerprints. Is there a reason why it uses the original method instead of the modified?

I made these changes (still uncleaned, just a proof-of-concept):
```diff
--- a/src/main/kotlin/app/revanced/util/patch/AbstractTransformInstructionsPatch.kt
+++ b/src/main/kotlin/app/revanced/util/patch/AbstractTransformInstructionsPatch.kt
@@ -24,10 +24,10 @@ internal abstract class AbstractTransformInstructionsPatch<T> : BytecodePatch()
     override fun execute(context: BytecodeContext): PatchResult {
         // Find all instructions
         buildMap {
-            context.classes.forEach { classDef ->
+            context.classes.map { context.proxy(it).mutableClass }.forEach { classDef ->
                 classDef.methods.let { methods ->
                     buildMap methodList@{
-                        methods.forEach methods@{ method ->
+                        methods.map(classDef::findMutableMethodOf).forEach methods@{ method ->
                             with(method.implementation?.instructions ?: return@methods) {
                                 ArrayDeque<T>().also { patchIndices ->
                                     this.forEachIndexed { index, instruction ->
@@ -50,9 +50,8 @@ internal abstract class AbstractTransformInstructionsPatch<T> : BytecodePatch()
             // And finally transform the instructions...
             with(context.proxy(classDef).mutableClass) {
                 methods.forEach { (method, patches) ->
-                    val mutableMethod = findMutableMethodOf(method)
                     while (!patches.isEmpty()) {
-                        transform(mutableMethod, patches.removeLast())
+                        transform(method, patches.removeLast())
                     }
                 }
             }
```

With these changes, it seems to work. But applying the patches takes longer now:
Before:
```
real    0m59.056s
user    1m44.309s
sys     0m3.500s
```

After:
```
real    3m57.534s
user    4m48.866s
sys     0m4.342s
```

Is there some way to speed it up?

**Comment by oSumAtrIX** [comment]
> Is there a reason it uses the original method instead of the modified one?

I don't quite remember, I think it was an implementation limitation or a choice.

> With these changes, it seems to work. But applying the patches takes longer now:

The changes you applied to cause an immense and unpractical overhead as it creates a mutable copy of all classes. A better approach would be first to locate the methods subject to be modified, then create mutable copies from them, then find the indices of the instructions to modify once again. This is not a good solution and should not be implemented this way; it is merely an example of optimizing and effectively working around this issue. Instead, a proper solution to the issue with fingerprints and shared mutability should be found.

**Comment by Linus789** [comment]
> A better approach would be first to locate the methods subject to be modified, then create mutable copies from them, then find the indices of the instructions to modify once again.

Good idea! The time has now been reduced to
```
real    0m58.928s
user    1m44.944s
sys     0m3.672s
```
which is as good as before.

> This is not a good solution and should not be implemented this way. Instead, a proper solution to the issue with fingerprints and shared mutability should be found.

Sure. Would you mind merging a PR that fixes this issue until a better solution has been found?

**Comment by oSumAtrIX** [comment]
> Sure. Would you mind merging a PR that fixes this issue until a better solution has been found?

If possible, open a tracking issue for this, referencing anything relevant.

**Comment by elim-garak-dev** [comment]
"fix(spoof-wifi-connection): use updated instruction indices" patch solve my problem. No more crash.
